### ISSUE TITLE
Different behavior of pandoc's Lua in Win and Linux (UTF8 and patterns)

### ISSUE THREAD

**Issue by kysko** [jgm/pandoc#6077]
**TLDR;**
I've encountered a difference of behavior between pandoc's Lua on Windows and Linux distro, when it comes to UTF8 and patterns.
Lua standalone on Windows by itself doesn't seem to exhibit this problem (no extensive test though), so this might not be related to usual Windows console problems with utf8.

**In details:**

(with Windows 10 x64, Linux Mint 19.3 x64 in a VM, pandoc 2.9.1.1 recent Nigthlies)

We already know one should be careful with UTF8 texts and Lua's patterns in string functions, because they usually act on characters as single bytes.
Before using an external module for advanced utf8 handling, I'd still like to know the reason for the following difference if possible.

Consider the following simple text (test.md)

```
Hé!
```
and the Lua filter:

```
-- test.lua
function Str(el)
  local a = el.text

  local s,e,cap = a:find("(é)") -- capture "é"
  local str = "character " .. cap .. " found at " .. s .. " and " .. e .. "\n"

  s,e,cap = a:find("(%p)")    -- capture first punctuation
  str = str .. "punctuation found at " .. s .. " and " .. e
  return pandoc.Str(str)
end
```
Using `pandoc test.md -L test.lua -o result.md` on a linux distro, we get the correct result:

```
character é found at 2 and 3
punctuation found at 4 and 4
```
("at 4 and 4" means "only at 4")

but on Windows, we get:

```
character é found at 2 and 3
punctuation found at 3 and 3
```

=> So the embedded Lua in pandoc found the first punctuation at different places: the correct one on linux distro, *but* on the second byte of "é" on Windows.

I thought this was *again* a problem with Windows and its command line... but then I used a native Lua on Windows10 (self-compiled 5.3.5)* on the following script (mostly same as above):

```
-- testnative.lua
local a="Hé!"
local s,e,cap = a:find("(é)")
local str = "character " .. cap .. " found at " .. s .. " and " .. e .. "\n"

s,e,cap = a:find("(%p)")
str = str .. "punctuation found at " .. s .. " and " .. e
print(str)
```

And I get the _correct_ result in the console:

```
> chcp 65001
> lua.exe testnative.lua
character é found at 2 and 3
punctuation found at 4 and 4
```
(the `chcp 65001` above is only for output display in the Win console; using it or not with pandoc in the previous example doesn't change the result.)

*(/Edit: `luatex.exe --luaonly testnative.lua` from TexLive2019 gives same correct result)

**_Question:_** So is Lua embedded differently in Windows' pandoc? Is this a bug, or is it expected?

_Additional detail:_

We can get some info with following script:

```
-- test0.lua
function Str(el)
  local a, t = el.text, {}

  for p, c in utf8.codes(a) do
    table.insert(t, "(p=" .. p .. ", c=" .. c .. ")")
  end
  local str = "codepoint sequence: " .. table.concat(t, ", ")

  t = {}
  for i=1, a:len() do
    table.insert(t, "(p=" .. i .. ", b=" .. string.byte(a,i) .. ")")
  end
  str = str .. "\n" .. "byte sequence     : " .. table.concat(t, ", ")
  return pandoc.Str(str)
end
```
which give (both Win and Linux) (`pandoc test.md -L test0.lua -o result0.md`):

```
codepoint sequence: (p=1, c=72), (p=2, c=233), (p=4, c=33)
byte sequence     : (p=1, b=72), (p=2, b=195), (p=3, b=169), (p=4, b=33)
```
As expected the character `é` is represented by two bytes in UTF8 (`\195\169`, or `\xC3\xA9`).

**Comment by jgm** [comment]
@tarleb may be able to shed light on this.

**Comment by kysko** [comment]
Trying again after some sleep...

Using
```
assert(os.setlocale("C"))
-- or assert(os.setlocale("C", "ctype"))
```
seems to solve this particular situation. Will have to test consequences.

By default, `os.locale()` from pandoc's Lua gave me
```
LC_COLLATE=C;LC_CTYPE=English_Canada.1252;LC_MONETARY=C;LC_NUMERIC=C;LC_TIME=C
```
(The native Lua executables (the one I compiled, and the one from luatex in TexLive2019) gave me `C` for all.)

Trying
`assert(os.setlocale("en_CA.UTF-8", "ctype"))`,
`assert(os.setlocale("en_CA.UTF-8"))`,
`assert(os.setlocale("English_Canada.UTF-8", "ctype"))`
or `assert(os.setlocale("en_US.UTF-8", "ctype"))`
didn't work, throws error.

**Comment by tarleb** [comment]
That's interesting, thanks for the analysis. I'm not really sure how we should go about fixing this. It would be great to have a Haskell UTF-8 implementation of the Lua pattern matcher, but that seems like a lot of work.

**Comment by kysko** [comment]
I was more worried about consistency for now, and `os.setlocale('C', 'ctype')` might bring that.

If I can match at least on punctuations, I can split without worries (with the help of lua's utf8 string library, see below).

Since the 'native' lua consoles I have (on Windows) show`LC_CTYPE=C`, does that mean you had to set some flag at compilation to set it to the users' OS's locale? i.e. is it a deliberate choice to make the interpreter work better with pandoc?

Or would `LC_CTYPE=C` be more "neutral"/OS agnostic?
(I might guess that `pandoc.read(...)` helper function might need the OS's locale though...)

> It would be great to have a Haskell UTF-8 implementation of the Lua pattern matcher, but that seems like a lot of work.

I'm sure many would like better UTF-8 support natively in Lua in general, at least for the main pattern-matching functions.

I've stumbled on a 2014 mailing list discussion [UTF-8 patterns in Lua 5.3](http://lua-users.org/lists/lua-l/2014-04/msg00820.html) by some "Hisham" who tried to remedy the situation. [Four years later](http://lua-users.org/lists/lua-l/2018-07/msg00295.html) he explained what "turned me off about the idea".

The later was part of a discussion on the [Lua utf8 library](http://lua-users.org/lists/lua-l/2018-07/msg00286.html), itself a continuation of a much longer discussion about [Lua and UTF8](http://lua-users.org/lists/lua-l/2018-07/msg00198.html) ...
So, no, I wouldn't expect a quick solution as a Haskell implementation!

At least someone gave a quick version of what a [utf8.sub function](http://lua-users.org/lists/lua-l/2014-04/msg00590.html) can look like only using Lua 5.3 utf8 library. Not a pattern matching, but useful.

**Comment by tarleb** [comment]
> does that mean you had to set some flag at compilation to set it to the users' OS's locale?

There is none that I'm aware of. I'll check again.

It looks like we have two options: we could either set the locale to a default value before calling Lua, or we could document the current situation to be clear that the locale can affect pattern matching. Using a default value seems more in line with how we usually handle things.

> So, no, I wouldn't expect a quick solution as a Haskell implementation!

I thought about submitting a couple of Lua related proposals for the [Haskell Summer of Code](https://summer.haskell.org). Lua-like string matching on `Text` might be an interesting summer project.

---

Slightly off-topic comment triggered by seeing Hisham mentioned: We met last year and discussed whether inclusion of a luarocks like system into pandoc could make sense. There is a lot of work being done to modularize luarocks, so that might become an option in the future. It would be great if one could pull-in a good UTF-8 library just through pandoc; but for now, we'll have to make do with system-wide Lua setups.

**Comment by kysko** [comment]
> either set the locale to a default value before calling Lua, or we could document the current situation (...)

Either case would have to be documented I think. What would be the default ctype value? `C`?
I'm still not totally sure what `LC_CTYPE=C` means/entails for Lua. Is it OS independent?

> Lua-like string matching on [Text]

A project that would be in good hands if accepted 👍 

> Slightly off-topic comment (...)

After your comment, I just searched and realized the Hisham I linked was the maintainer of LuaRocks... didn't know that.

**Comment by kysko** [comment]
Still a few more comments on the situation...

For `LC_CSTYPE=C` (or the default `LC_CSTYPE=xyz.UTF-8` on Linux distro), the punctuation range `%p` is: `\33-\47, \58-\64, \91-\96, \123-\126`, that is within the normal ASCII limits.

For `LC_CSTYPE=xyz.1252`, the default on Windows (western countries), the punctuation range for `%p` is as above, but *augmented* with the following range: `\130, \132-\135, \137, \139, \145-\151, \155, \161-\191, \215, \247` (still 1 byte though), thus including the *"extended ASCII"* (or ANSI, or whatever in this mess).

In particular, in Windows, for my original example, the "punctuation" `\169` (copyright symbol) was matched with the second byte of `é == \195\169`.
Setting `LC_CSTYPE=C` was thus just a way to restrict the range of `%p` in this case.

Anyway, this is an addition to the usual warnings about string manipulations with patterns in Lua.
I might not change `LC_CSTYPE` after all, and use an explicit pattern rather than the class `%p` (or other classes), and check whether the captured bytes are part of a double-byte character, etc... Or just use a module/library that does this in the background.

What is certain is that the default value of `LC_CSTYPE` is different for pandoc's embedded Lua on Windows and *nix, which is normal as it reflects the system it is installed on.
But this affects the ranges of character classes (or at least `%p`, didn't do an extended survey) and one must be careful for the portability of Lua filters.

**Comment by tarleb** [comment]
We now have an entry for this in the list of common pitfalls. Thanks again for tracking this down @kysko, I learned something new this way.

**Comment by jgm** [comment]
@tarleb is there a way to force LC_CSTYPE for the Lua pandoc is using, to remove this potential pitfall?

**Comment by tarleb** [comment]
I believe it might be enough to set, and later reset, the `LC_CTYPE` environment variable to `C` before running Lua. That might be problematic though, as it would also affect all programs called from Lua filters, with difficult to predict consequences.
